From c9976311cf4255037216fa80131c0bd2e661e570 Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Mon, 6 Apr 2026 06:53:27 +0000
Subject: [PATCH] can: j1939: j1939_session_new(): fix skb reference counting

jira VULN-167655
cve CVE-2024-56645
commit-author Dmitry Antipov <dmantipov@yandex.ru>
commit a8c695005bfe6569acd73d777ca298ddddd66105

Since j1939_session_skb_queue() does an extra skb_get() for each new
skb, do the same for the initial one in j1939_session_new() to avoid
refcount underflow.

	Reported-by: syzbot+d4e8dc385d9258220c31@syzkaller.appspotmail.com
Closes: https://syzkaller.appspot.com/bug?extid=d4e8dc385d9258220c31
Fixes: 9d71dd0c7009 ("can: add support of SAE J1939 protocol")
	Signed-off-by: Dmitry Antipov <dmantipov@yandex.ru>
	Tested-by: Oleksij Rempel <o.rempel@pengutronix.de>
	Acked-by: Oleksij Rempel <o.rempel@pengutronix.de>
Link: https://patch.msgid.link/20241105094823.2403806-1-dmantipov@yandex.ru
[mkl: clean up commit message]
	Signed-off-by: Marc Kleine-Budde <mkl@pengutronix.de>
(cherry picked from commit a8c695005bfe6569acd73d777ca298ddddd66105)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 net/can/j1939/transport.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/net/can/j1939/transport.c b/net/can/j1939/transport.c
index 846e0717b406..06a819a23e67 100644
--- a/net/can/j1939/transport.c
+++ b/net/can/j1939/transport.c
@@ -1483,7 +1483,7 @@ static struct j1939_session *j1939_session_new(struct j1939_priv *priv,
 	session->state = J1939_SESSION_NEW;
 
 	skb_queue_head_init(&session->skb_queue);
-	skb_queue_tail(&session->skb_queue, skb);
+	skb_queue_tail(&session->skb_queue, skb_get(skb));
 
 	skcb = j1939_skb_to_cb(skb);
 	memcpy(&session->skcb, skcb, sizeof(session->skcb));
-- 
2.47.3

